### PR TITLE
major updates

### PR DESCRIPTION
**Pull request by lostmygithubaccount**
done:

- rename notebooks
- run formatting
- follow format of other notebooks
- separate out single training vs hyperdrive training 
- edits to the dockerfile 
- estimator to scriptrunconfig

to do:

- get working
- switch logging to mlflow 

current issue:

running into a cuda issue:

```
The experiment failed. Finalizing run...
[2020-09-30T19:49:13.076075] TimeoutHandler __init__
[2020-09-30T19:49:13.076157] TimeoutHandler __enter__
Cleaning up all outstanding Run operations, waiting 300.0 seconds
3 items cleaning up...
Cleanup took 0.4709360599517822 seconds
[2020-09-30T19:49:13.668172] TimeoutHandler __exit__
Enter __exit__ of DatasetContextManager
Unmounting /tmp/tmpwch2sl2e.
Finishing unmounting /tmp/tmpwch2sl2e.
Exit __exit__ of DatasetContextManager
Traceback (most recent call last):
  File "train.py", line 127, in <module>
    main()
  File "train.py", line 101, in main
    X_train, X_test, y_train, y_test, _ = azure_ml.split_data(X, y, random_state=i_train_fold)
  File "/mnt/batch/tasks/shared/LS_root/jobs/default/azureml/rapids-airline-example_1601494194_1354c796/mounts/workspaceblobstore/azureml/rapids-airline-example_1601494194_1354c796/rapids_csp_azure.py", line 280, in split_data
    random_state = random_state)
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cuml/preprocessing/model_selection.py", line 375, in train_test_split
    X = X.iloc[idxs]
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/indexing.py", line 180, in __getitem__
    return self._getitem_tuple_arg(arg)
  File "/opt/conda/envs/rapids/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/indexing.py", line 422, in _getitem_tuple_arg
    df = columns_df._gather(arg[0])
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/frame.py", line 488, in _gather
    self, as_column(gather_map), keep_index=keep_index
  File "cudf/_lib/copying.pyx", line 145, in cudf._lib.copying.gather
RuntimeError: after reduction step 1: cudaErrorInvalidDeviceFunction: invalid device function


Error occurred: AzureMLCompute job failed.
JobFailed: Submitted script failed with a non-zero exit code; see the driver log file for details.
```

### CODE REVIEW

**Comment by lostmygithubaccount** [comment]
@zronaghi can you advise on the error? not sure how to proceed

**Comment by zronaghi** [comment]
@lostmygithubaccount which GPUs/compute instance did you use for this test?

**Comment by lostmygithubaccount** [comment]
`STANDARD_NC12` which has K80s

**Comment by zronaghi** [comment]
Ok, so RAPIDS is supported on Pascal and later GPU, could you test with VMs that have P40, P100 or V100s? I've added a mention in the notebook as well

**Comment by lostmygithubaccount** [comment]
will upgrade to NC_12_V2 (P100s) and try

**Comment by lostmygithubaccount** [comment]
I got it running on V100s - can you do a sanity check that it's using both GPUs to train the model correctly? 

There are a few more things to do - I should be able to get them done today - then if you could review and merge this PR we can then merge your PR back into Azure

todo:

- [x] reduce training time of runs (<8 mins) 
- [x] multi-gpu setting added to argparse
- [x] switch to mlflow logging 
- [x] persist dockerfile in environments 
- [ ] add description metadata
- [ ] run readme.py

**Comment by lostmygithubaccount** [comment]
@zronaghi new issue - any ideas? - trying out multi-GPU fails with this: 

```
---->>>> cuDF version <<<<----
 0.15.0

---->>>> cuML version <<<<----
 0.15.0

> RapidsCloudML
	Compute, Data , Model, Cloud types ('multi-GPU', 'Parquet', 'RandomForest', 'Azure')

	Multi-GPU selected

	Client information <Client: 'tcp://127.0.0.1:46225' processes=2 threads=2, memory=236.67 GB>
multi-GPU

> Loading dataset from /tmp/tmpszqbfhn9/part*.parquet

	GPU read

	Reading using dask_cudf
Starting the daemon thread to refresh tokens in background for process with pid = 284


[2020-09-30T23:51:12.772482] The experiment failed. Finalizing run...
[2020-09-30T23:51:12.772659] FinalizingInRunHistory is not called
Cleaning up all outstanding Run operations, waiting 300.0 seconds
3 items cleaning up...
Cleanup took 1.154808521270752 seconds
Enter __exit__ of DatasetContextManager
Unmounting /tmp/tmpszqbfhn9.
Finishing unmounting /tmp/tmpszqbfhn9.
Exit __exit__ of DatasetContextManager
Traceback (most recent call last):
  File "train.py", line 179, in <module>
    main()
  File "train.py", line 116, in main
    y_label="ArrDel15",
  File "/mnt/batch/tasks/shared/LS_root/jobs/default/azureml/rapids-airline-example_1601509638_65820726/mounts/workspaceblobstore/azureml/rapids-airline-example_1601509638_65820726/rapids_csp_azure.py", line 219, in load_data
    target_filename, columns=col_labels
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cudf/io/parquet.py", line 189, in read_parquet
    **kwargs,
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/io/parquet/core.py", line 239, in read_parquet
    **kwargs,
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cudf/io/parquet.py", line 17, in read_metadata
    meta, stats, parts, index = ArrowEngine.read_metadata(*args, **kwargs)
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/io/parquet/arrow.py", line 656, in read_metadata
    paths, fs, split_row_groups, gather_statistics, filters, dataset_kwargs
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/io/parquet/arrow.py", line 166, in _gather_metadata
    dataset, base, fns = _get_dataset_object(paths, fs, filters, dataset_kwargs)
  File "/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/io/parquet/arrow.py", line 136, in _get_dataset_object
    elif fs.isdir(paths[0]):
IndexError: list index out of range

[2020-09-30T23:51:15.774070] Finished context manager injector with Exception.

Error occurred: AzureMLCompute job failed.
JobFailed: Submitted script failed with a non-zero exit code; see the driver log file for details.
```

**Comment by lostmygithubaccount** [comment]
ah, because I'm using the smaller dataset? 

edit: seems to be issues regardless using multi-GPU, switching back to single-GPU for now

**Comment by lostmygithubaccount** [comment]
pushed my changes - can you help with mutli-GPU training debugging and also logging the models?

**Comment by zronaghi** [comment]
Yes the multi-GPU version expects multiple parquet files, did you see the same issue with `airline-10years` dataset? And you mean logging the models after training?

**Comment by lostmygithubaccount** [comment]
yes, logging the models after training. I do not see the same issue with `airline-10years` dataset, but there is a new issue - CUDA memory runs out and it also throws an error that `trained_model` is referenced before assignment

**Comment by lostmygithubaccount** [comment]
log file: 

```
[2020-10-01T01:11:04.939109] Entering context manager injector.
[context_manager_injector.py] Command line Options: Namespace(inject=['ProjectPythonPath:context_managers.ProjectPythonPath', 'Dataset:context_managers.Datasets', 'RunHistory:context_managers.RunHistory', 'TrackUserError:context_managers.TrackUserError', 'UserExceptions:context_managers.UserExceptions'], invocation=['train.py', '--data_dir', 'DatasetConsumptionConfig:input_6a85608e', '--n_bins', '32', '--compute', 'multi-GPU', '--cv-folds', '1'])
This is an MPI job. Rank:0
Initialize DatasetContextManager.
Starting the daemon thread to refresh tokens in background for process with pid = 279
Set Dataset input_6a85608e's target path to /tmp/tmpgmcls4e9
Enter __enter__ of DatasetContextManager
SDK version: azureml-core==1.15.0 azureml-dataprep==2.3.0. Session id: eba7fc22-707a-4976-8eaa-816bf7e94a20. Run id: rapids-airline-multi-example_1601514635_4bf8dba2.
Processing 'input_6a85608e'.
Could not refresh EngineServer credentials in rslex: No Lariat Runtime Environment is active, please initialize an Environment.
Could not refresh EngineServer credentials in rslex: No Lariat Runtime Environment is active, please initialize an Environment.
Processing dataset FileDataset
{
  "source": [
    "https://airlinedataset.blob.core.windows.net/airline-10years/*"
  ],
  "definition": [
    "GetFiles"
  ],
  "registration": {
    "id": "a13fb474-eb17-43e7-9432-da7a83bef4b3",
    "name": null,
    "version": null,
    "workspace": "Workspace.create(name='default', subscription_id='6560575d-fa06-4e7d-95fb-f962e74efd7a', resource_group='azureml-examples')"
  }
}
Mounting input_6a85608e to /tmp/tmpgmcls4e9.
Could not refresh EngineServer credentials in rslex: No Lariat Runtime Environment is active, please initialize an Environment.
Could not refresh EngineServer credentials in rslex: No Lariat Runtime Environment is active, please initialize an Environment.
Mounted input_6a85608e to /tmp/tmpgmcls4e9 as folder.
Exit __enter__ of DatasetContextManager
Entering Run History Context Manager.
Current directory:  /mnt/batch/tasks/shared/LS_root/jobs/default/azureml/rapids-airline-multi-example_1601514635_4bf8dba2/mounts/workspaceblobstore/azureml/rapids-airline-multi-example_1601514635_4bf8dba2
Preparing to call script [ train.py ] with arguments: ['--data_dir', '$input_6a85608e', '--n_bins', '32', '--compute', 'multi-GPU', '--cv-folds', '1']
After variable expansion, calling script [ train.py ] with arguments: ['--data_dir', '/tmp/tmpgmcls4e9', '--n_bins', '32', '--compute', 'multi-GPU', '--cv-folds', '1']

Script type = None

---->>>> cuDF version <<<<----
 0.15.0

---->>>> cuML version <<<<----
 0.15.0

> RapidsCloudML
	Compute, Data , Model, Cloud types ('multi-GPU', 'Parquet', 'RandomForest', 'Azure')

	Multi-GPU selected
/opt/conda/envs/rapids/lib/python3.7/site-packages/distributed/node.py:155: UserWarning: Port 8787 is already in use.
Perhaps you already have a cluster running?
Hosting the HTTP server on port 45857 instead
  http_address["port"], self.http_server.port

	Client information <Client: 'tcp://127.0.0.1:43017' processes=2 threads=2, memory=236.67 GB>
multi-GPU

> Loading dataset from /tmp/tmpgmcls4e9/part*.parquet

	GPU read

	Reading using dask_cudf
cudf/utilities/bit.hpp(19): warning: cassert: [jitify] File not found
cudf/fixed_point/fixed_point.hpp(27): warning: cassert: [jitify] File not found

	Ingestion completed in 3.732228081998983

	Dataset descriptors: (Delayed('int-b18bf357-d836-4ac7-9f28-645e53b6b5af'), 15)
	Flight_Number_Reporting_Airline    float32
Year                               float32
Quarter                            float32
Month                              float32
DayOfWeek                          float32
DOT_ID_Reporting_Airline           float32
OriginCityMarketID                 float32
DestCityMarketID                   float32
DepTime                            float32
DepDelay                           float32
DepDel15                           float32
ArrDel15                             int32
ArrDelay                           float32
AirTime                            float32
Distance                           float32
dtype: object

---->>>> Training using GPUs <<<<----


 CV fold 0 of 1


> Splitting train and test data

	X_train shape and type(Delayed('int-be7406a5-7e4a-4e16-8da8-81ea4d98220b'), 13) <class 'dask_cudf.core.DataFrame'>

	Split completed in 0.004595635000441689

> Training RandomForest estimator w/ hyper-params

	Fitting multi-GPU daskRF
/opt/conda/envs/rapids/lib/python3.7/site-packages/cuml/dask/ensemble/randomforestclassifier.py:158: UserWarning: For reproducible results in Random Forest Classifier or for almost reproducible results in Random Forest Regressor, n_streams==1 is recommended. If n_streams is > 1, results may vary due to stream/thread timing differences, even when random_seed is set
  **kwargs
/opt/conda/envs/rapids/lib/python3.7/site-packages/cuml/dask/ensemble/randomforestclassifier.py:158: UserWarning: For reproducible results in Random Forest Classifier or for almost reproducible results in Random Forest Regressor, n_streams==1 is recommended. If n_streams is > 1, results may vary due to stream/thread timing differences, even when random_seed is set
  **kwargs
distributed.worker - WARNING -  Compute Failed
Function:  _func_fit
args:      (RandomForestClassifier(split_criterion=0), [(        AirTime  DOT_ID_Reporting_Airline  ...  Quarter    Year
index                                      ...                 
0         148.0                   19977.0  ...      1.0  2018.0
1          98.0                   19977.0  ...      1.0  2018.0
2         190.0                   19977.0  ...      1.0  2018.0
3         212.0                   19977.0  ...      1.0  2018.0
4         112.0                   19977.0  ...      1.0  2018.0
...         ...                       ...  ...      ...     ...
570110     75.0                   19790.0  ...      1.0  2018.0
570113     63.0                   19790.0  ...      1.0  2018.0
570114    126.0                   19790.0  ...      1.0  2018.0
570116     91.0                   19790.0  ...      1.0  2018.0
570117    279.0                   19790.0  ...      1.0  2018.0

[455919 rows x 13 columns], index
0         0
1         0
2         0
3         0
4         0
         ..
570110    1
570
kwargs:    {}
Exception: MemoryError('std::bad_alloc: CUDA error at: ../include/rmm/mr/device/cuda_memory_resource.hpp:68: cudaErrorMemoryAllocation out of memory')

distributed.worker - WARNING -  Compute Failed
Function:  _func_fit
args:      (RandomForestClassifier(split_criterion=0), [(        AirTime  DOT_ID_Reporting_Airline  ...  Quarter    Year
index                                      ...                 
0          37.0                   20304.0  ...      1.0  2019.0
3          80.0                   20304.0  ...      1.0  2019.0
4         127.0                   20304.0  ...      1.0  2019.0
5         110.0                   20304.0  ...      1.0  2019.0
6          72.0                   20304.0  ...      1.0  2019.0
...         ...                       ...  ...      ...     ...
583977     41.0                   19690.0  ...      1.0  2019.0
583978     33.0                   19690.0  ...      1.0  2019.0
583982     22.0                   19690.0  ...      1.0  2019.0
583983     34.0                   19690.0  ...      1.0  2019.0
583984     30.0                   19690.0  ...      1.0  2019.0

[467159 rows x 13 columns], index
0         0
3         0
4         0
5         0
6         0
         ..
583977    0
583
kwargs:    {}
Exception: MemoryError('std::bad_alloc: CUDA error at: ../include/rmm/mr/device/cuda_memory_resource.hpp:68: cudaErrorMemoryAllocation out of memory')



! Error during fit 2 of 2 worker jobs failed: std::bad_alloc: CUDA error at: ../include/rmm/mr/device/cuda_memory_resource.hpp:68: cudaErrorMemoryAllocation out of memory, std::bad_alloc: CUDA error at: ../include/rmm/mr/device/cuda_memory_resource.hpp:68: cudaErrorMemoryAllocation out of memory


!error during model training: local variable 'trained_model' referenced before assignment

	Finished training in 0.0000 s
Starting the daemon thread to refresh tokens in background for process with pid = 279


[2020-10-01T01:13:43.020696] The experiment failed. Finalizing run...
[2020-10-01T01:13:43.020852] FinalizingInRunHistory is not called
Cleaning up all outstanding Run operations, waiting 300.0 seconds
3 items cleaning up...
Cleanup took 5.351221323013306 seconds
Enter __exit__ of DatasetContextManager
Unmounting /tmp/tmpgmcls4e9.
Finishing unmounting /tmp/tmpgmcls4e9.
Exit __exit__ of DatasetContextManager
Traceback (most recent call last):
  File "train.py", line 179, in <module>
    main()
  File "train.py", line 149, in main
    X_train, y_train, model_params
  File "/mnt/batch/tasks/shared/LS_root/jobs/default/azureml/rapids-airline-multi-example_1601514635_4bf8dba2/mounts/workspaceblobstore/azureml/rapids-airline-multi-example_1601514635_4bf8dba2/rapids_csp_azure.py", line 344, in train_model
    return trained_model, training_time
UnboundLocalError: local variable 'trained_model' referenced before assignment

[2020-10-01T01:13:53.706456] Finished context manager injector with Exception.
```

**Comment by zronaghi** [comment]
I tested these with 4x V100s, would you like to keep 2 GPUs the minimum for multi-GPU examples? In that case I can modify the dataset to use fewer files. (and the second error is probably also caused by the OOM)

Btw these are great that you've separated the notebook examples, definitely will be easier to follow :)

**Comment by lostmygithubaccount** [comment]
do you know how to properly run on 4x V100s? I'm trying to use `STANDARD_NC12_V3` which has 2 V100s per node, then running 2 nodes - see the `train-airplaines-multi.ipynb`. I'm using an MpiConfiguration, which I don't think is really what should be used for Dask. 

do I need to use `STANDARD_NC24_V3` and a single node? is this designed to work on multinode compute targets? 

cc: @mx-iao

**Comment by zronaghi** [comment]
Correct, these are setup to work on single node with multiple GPUs for now (we're working on extending these examples to multi-node multi-GPU with Dask), I thought `Standard_NC24s_v3` always allocates 4 V100s on a node.

**Comment by lostmygithubaccount** [comment]
yes, the `STANDARD_NC12_V3` I was using allocates 2 V100s on a node.

I am updating to use a NC24, will test and push final changes soon

**Comment by lostmygithubaccount** [comment]
awesome, all 3 notebooks work now - I'm going to update the repo to create/use multiple GPU targets, but I'll do that separately (https://github.com/Azure/azureml-examples/issues/73).

I think this PR is ready for you to review/merge.

**Comment by zronaghi** [comment]
Awesome, thank you! I'll review it now

File: environments/rapids-example.dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM rapidsai/rapidsai:0.15-cuda10.2-runtime-ubuntu18.04-py3.7
+RUN apt-get update && \
+apt-get install -y fuse && \
+apt-get install -y openmpi-bin openmpi-common openssh-client openssh-server libopenmpi2 libopenmpi-dev && \
```
Comment (lostmygithubaccount) [inline environments/rapids-example.dockerfile]: this can probably be removed

File: setup.ipynb [hunk 2]
```diff
@@ -69,7 +69,7 @@
     "\n",
     "if ct_name not in ws.compute_targets:\n",
     "    compute_config = AmlCompute.provisioning_configuration(\n",
-    "        vm_size=\"STANDARD_NC12\",\n",
+    "        vm_size=\"STANDARD_NC12_V2\",\n",
```
Comment (lostmygithubaccount) [inline setup.ipynb:72]: this needs to be updated again, will do separately

File: notebooks/rapids/train-airlines-hyperdrive.ipynb [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from azureml.core import Workspace\n",
+    "\n",
+    "ws = Workspace.from_config(\"~/code/default.json\")\n",
```
Comment (zronaghi) [inline notebooks/rapids/train-airlines-hyperdrive.ipynb]: Probably can remove this path to config

File: notebooks/rapids/train-airlines-multi.ipynb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from azureml.core import Workspace\n",
+    "\n",
+    "ws = Workspace.from_config(\"~/code/default.json\")\n",
```
Comment (zronaghi) [inline notebooks/rapids/train-airlines-multi.ipynb]: Same here

**Pull request merged**